### DISCUSSION TITLE
Support three-state bool for CheckBox (and TreeView)

### DISCUSSION THREAD

**Discussion by KieranMaclagan** [enisn/UraniumUI#228] [Ideas]
An enhancement to the TreeView would be to have partially selected items have an indeterminate `null` state. Right now it returns `true`. Ideally through changing IsChecked from a `bool` to a `bool?`. This would let us easily identify if a node is actually selected instead of some of it's children. Another scenario this could be useful is for a "Select All" option that can have partial selection.

Perhaps instead of changing the existing CheckBox, since CheckBoxes are naturally either `true`/`false`, maybe adding a NullableCheckBox would be a better alternative?

![image](https://user-images.githubusercontent.com/10479497/216369362-d531827a-aa9e-4295-beec-b976bcbe35cf.png)

**Comment by enisn** [comment]
Thank you for discussing this case.

Firstly I'll explain the current situation and then we can discuss how it should be.

**Currently,** Indeterminate is not a state and it won't be checked until all of the children are checked. So, it doesn't have its own state. It represents the selection state of its children. When you check it, it'll automatically check all the children of itself. It stands for only the grouping of its children. In another hand, you can't understand if it's indeterminate or not programmatically without looking its children.

**In the best case,** there should be a state that can be used for binding. Since it's a simple behavior and it's not related to checkbox. The behavior might have a property that can be used to keep a state of the checkbox like the following:

```xml
<material:TreeView ItemsSource="{Binding Nodes}">
    <material:TreeView.ItemTemplate>
        <DataTemplate>
            <material:CheckBox Text="{Binding Name}">
                <material:CheckBox.Behaviors>
                    <material:TreeViewHierarchicalSelectBehavior State="{Binding State}" />
                </material:CheckBox.Behaviors>
            </material:CheckBox>
        </DataTemplate>
    </material:TreeView.ItemTemplate>
</material:TreeView>
``` 

So, state can represent 3 cases for `checked`, `indeterminate `, `notcheckted`.

I'll create an issue for that and include it in the next version (v2.4)

**Reply by KieranMaclagan** [reply]
Great idea, what you described would work well. Sounds good, thanks!
